pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1766,12 +1769,41 @@ func (c *Cloud) AttachDisk(diskName KubernetesVolumeID, nodeName types.NodeName,
 }
 
 // DetachDisk implements Volumes.DetachDisk
-func (c *Cloud) DetachDisk(diskName KubernetesVolumeID, nodeName types.NodeName) (string, error) {
+func (c *Cloud) DetachDisk(diskName KubernetesVolumeID, nodeName types.NodeName, nodeExists bool) (string, error) {
```

Comment:
It is not very clear for me to have nodeExists here. The function is to Detach disk from a node, I feel the logic of checking node exist or not should just inside of this function?

Reply:
Yeah it can be certainly changed. will do.

Reply:
So the main reason this additional boolean field has been introduced is because - `DiskIsAttached` check returns `false` (as in disk is detached) incorrectly when node is shutdown and volume is still attached to the node.  I am trying to fix that behaviour here.

The reason `DiskIsAsAttached` returns `false` even though Disk is still attached to a stopped node is because, in AWS a stopped node is not returned when fetching instance list, so AWS cloudprovider throws "Node Not Found error", whereas node is still there and volumes are still attached to it.

I am trying to fix this, without changing too much of internal of AWS cloudprovider's code. `DiskIsAttached` appears to be a standard function that is being used by all volume types. I am not 100% sure, if it makes sense to roll this into `DetachDisk` just for aws.

Reply:
The main purpose of this fix is that when reconciler tries to attach a disk to a node, but somehow the disk is already attached to another node (e.g., attaching operation timeout but eventually attached). 
Here it seems like you are also trying to fix another issue for DiskIsAttached. Could you make a separate PR for that so we can discuss more details.

Reply:
Yep, that is correct. But I think both are related. We couldn't have fixed detaching volumes from stopped nodes if We didn't had Dangling volume detection code. 

If you can explain what is bothering you with the fix, may be I will try to workaround or address that.

Reply:
It seems like DiskIsAsAttached() is not implemented correctly for AWS when node could not be found by cloud provider, so I feel instead of adding the bool value, we should fix that function instead.

Reply:
You are correct that `DiskIsAsAttached` does not do return correct value. But fixing it alone will not fix Detach from stopped nodes. For example,  even if we called `DetachDisk` then again https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1775 code will make `DetachDisk` return again without doing actual detach.

So at minimum I think we need to fix both `DiskIsAttached` as well as `DetachDisk` to detach disks correctly from stopped nodes. 

>But detaching volume from a stopped node does not rely on this detection? When a node is stopped, and removed from the cloud provider, reconciler will trigger detach, but detach will do nothing since it though the volume is already detached.

That is correct, but I was thinking also about the case of node getting stopped and c-m restarted or active master switched in HA environment. Both of those cases will cause node information to be lost and hence detach may not be attempted. So there are like 2 bugs present here. :(

I am thinking lets fix both `DetachDisk` and `DiskIsAttached` functions, so as they can correctly handle stopped instances.  This PR sort of fixes `DetachDisk` but leaves `DiskIsAttached` broken... I will fix that too and update the PR.

Reply:
If you prefer, I can fix  `DetachDisk` and `DiskIsAttached`  in a separate PR.

Reply:
I think having two PRs is better so it is easier to explain what we are trying to fix.